### PR TITLE
Vastly rewrite the bot

### PR DESCRIPTION
**Pull request by OoLunar**
This rewrites the bot to include the following features:

- A proper config system
- Proper logging
- A database for persistent storage
- Commands
- Support for multiple guilds and channels
- Performance improvements

The following still needs to be done:

- Help command
- `redirect list` and `ignore list`
- Docker improvements
- Github Action improvements

### CODE REVIEW

**Comment by inftord** [comment]
this is how over-engineering looks, kids

**Comment by OoLunar** [comment]
> this is how over-engineering looks, kids

Let me rephrase that in a more professional manner: This is how you create a maintainable application.

**Comment by inftord** [comment]
> > this is how over-engineering looks, kids
> 
> Let me rephrase that in a more professional manner: This is how you create a maintainable application.

:wires:

**Comment by OoLunar** [comment]
This PR is now ready for review. When merging, I suggest to squash the commits.

File: src/Program.cs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+public sealed class Program
+{
+    internal static readonly IServiceCollection Services;
+    internal static readonly IConfiguration Configuration;
+    internal static readonly Database Database;
+    internal static readonly ILoggerFactory LoggerFactory;
+    private static readonly ILogger<Program> Logger;
+
+    static Program()
```
Comment (Instellate) [inline src/Program.cs:26]: I prefer to be explicit with access modifier

Comment (Instellate) [reply src/Program.cs:26]: Ignore this. I did not know static constructors was a thing and thought this was some weird method

File: src/Database.cs [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+using Microsoft.Data.Sqlite;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+
+namespace NotifierRedirecter;
+
+public sealed class Database
+{
+    // Sqlite is thread safe by default, meaning it's safe to use the same connection across multiple threads: https://stackoverflow.com/a/39916593/10942966
+    private readonly SqliteConnection _connection;
```
Comment (Instellate) [inline src/Database.cs:14]: Why use `SqliteConnection` instead of `SqliteFactory`?

File: src/Program.cs (hunk 1 above)
Comment (OoLunar) [reply src/Program.cs:26]: Static constructors don't have modifiers.

File: src/Database.cs (hunk 2 above)
Comment (OoLunar) [reply src/Database.cs:14]: Because we're only using a single connection. The connection itself is thread safe and using a factory would not provide any benefits that the connection doesn't already provide.

File: src/Program.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    {
+        Configuration = new ConfigurationBuilder()
+            .AddJsonFile("config.json", true, true)
```
Comment (Instellate) [inline src/Program.cs:29]: I might be blind. But I don't see a example for a config.json. If a example config does not exist. I would prefer to have one!

File: src/Database.cs (hunk 2 above)
Comment (Instellate) [reply src/Database.cs:14]: Ah ic

**Review by inftord — APPROVED** [review]
LGTM

File: src/Program.cs (hunk 3 above)
Comment (OoLunar) [reply src/Program.cs:29]: [`res/config.json`](https://github.com/OoLunar/NotifierRedirecter/blob/main/res/config.json)

Comment (Instellate) [reply src/Program.cs:29]: Ah there. Yea I was blind!

**Comment by Instellate** [comment]
I will merge this and hope this works

**Comment by OoLunar** [comment]
For once, everything is actually tested. In theory nothing will go wrong.
